content/blog/2023-12-09-hcl-amc/index.md

```diff
@@ -0,0 +1,259 @@
+          %4 = arith.muli %1, %2 : i32
+          %5 = arith.addi %3, %4 : i32
+          amc.affine_store %5, %0#1[%arg3, %arg4] : !amc.static_port<16x16xi32, rw, 1>
+        } {hls.pipeline, hls.unroll = 1 : i64}
+      } {hls.unroll = 1 : i64}
+    } {hls.unroll = 1 : i64}
+    affine.for %arg3 = 0 to 16 {
+      affine.for %arg4 = 0 to 16 {
+        %1 = amc.affine_load %0#1[%arg3, %arg4] : !amc.static_port<16x16xi32, rw, 1>
+        affine.store %1, %arg2[%arg3, %arg4] : memref<16x16xi32>
+      } {hls.bufferize, hls.pipeline, hls.unroll = 1 : i64}
+    } {hls.unroll = 1 : i64}
+    return
+  }
+}
+```
+
+For more information on the motivation behind AMC and the dialect itself, you can visit the [Andrew Butt's blog post](https://www.cs.cornell.edu/courses/cs6120/2022sp/blog/banked-memory-compiler-backend-in-mlir/) from last year. As a last resort when debugging, the AMC backend has a `.dump_vcd()` method which outputs the waveforms from the hardware simulation:
+
+```python
+    f.dump_vcd("waveform.vcd")
+    # You can now take the debugging to your favorite waveform viewer
+```
+
+Writing other types of kernels is as easy as writing normal Python. For example, here are a couple of other kernels which are just a simple to get up and running:
+
+```python
+def fibonnaci(A: uint32[N]):
+    A[0] = 1
+    A[1] = 1
+    for i in range(2, N):
+        A[i] = A[i - 1] + A[i - 2]
+
+def vadd(A: uint32[N], B: uint32[N]) -> uint32[N]:
+    return A + B
+```
+
+Outside of what is shown here, the HCL DSL allows much more elaborate kernels and control over compute customizations, like loop tiling and reuse buffers. One again, you can get more information on HCL and its MLIR dialect by reading last years blog post by [Hongzheng Chen](https://www.cs.cornell.edu/courses/cs6120/2022sp/blog/hcl-mlir/).
+
+To conclude, we hope this example demonstrates the usefulness of the HCL frontend for high-level hardware design and further development of the AMC HLS compiler. As far as we are aware, the only other frameworks using Python to drive FPGA high-level synthesis are [LeFlow](https://arxiv.org/abs/1807.05317) and [PyLog](https://ieeexplore.ieee.org/document/9591456). However, neither of these efforts are using a homebrewed HLS compiler like us.
+
+## Tool flow
+
+Under the hood, the HCL frontend is automating all the interactions with other tools, IRs, and frameworks. Nonetheless, understanding the interactions of each component is important to understanding the novelty in our approach.
+
+<center>
+<img src="hcl_dependencies.png" alt="Diagram of build dependencies" title="HCL build dependencies" style="zoom:45%;">
+</center>
+
+The top row of dependencies are C++ codebases linked together as static libraries. We enter this library with input MLIR generated from HCL. On the backend, we emit Calyx which is lowered to Verilog by a separate [Calyx compiler](https://github.com/cucapra/calyx) that is written in Rust.
```

Comment:
Just noting here that, while this overview is helpful, it doesn't directly answer the question I have above about the first step of the translation from Python. No need to change this here if you explain that step somewhere above.